### PR TITLE
[Core] Masternode collateral index

### PR DESCRIPTION
**Pull request by random-zebra**
Currently we check whether a masternode collateral is spent, in `CMasternode::Check()`, which is called in a time-based polling fashion, looking for the collateral outpoint in the coins view cache (and countless other times on-demand, when we want an up-to-date status).
This is highly expensive, and inconsistent (peers check masternodes at different times, thus have slightly different masternode states).

Collaterals, as all coins/outpoints, can only be spent inside blocks.

So we should do this check only when connecting new blocks, not *every x seconds*, or **every time that we need to count the number of enabled masternodes**, just to name the most glaring example (`CMasternodeMan::CountEnabled`).
This can be done very efficiently if we change the data-structure to store them inside the manager, from a vector, to a map, indexed by collateral outpoints.

----
Second part, to be addressed in a follow-up PR:

After this change, `CMasternode::Check()` includes only the ping-checks, which are so fast that can be done in "real-time", thus completely removing the need to cache the `activeState` inside the CMasternode class, and to call `Check()` every time that we want to update it.

### CODE REVIEW

File: src/qt/pivx/mnmodel.cpp [hunk 1]
```diff
@@ -31,7 +31,7 @@ void MNModel::updateMNList()
         if (!pmn) {
             pmn = new CMasternode();
             pmn->vin = CTxIn();
-            pmn->activeState = CMasternode::MASTERNODE_MISSING;
+            pmn->activeState = CMasternode::MASTERNODE_REMOVE;
```
Comment (furszy) [inline src/qt/pivx/mnmodel.cpp:34]: This isn't a "remove", this is a "not found". It happens when the entry is in the config file only, without broadcasting the start message.
The `ColumnIndex::STATUS` data is already contemplating this actually, we can leave it as nullptr and remove this whole if block. No need to create an empty masternode here.

File: src/masternode.h [hunk 2]
```diff
@@ -201,9 +195,6 @@ class CMasternode : public CSignedMessage
         READWRITE(protocolVersion);
         READWRITE(activeState);
         READWRITE(lastPing);
-        READWRITE(unitTest);
-        READWRITE(allowFreeTx);
-        READWRITE(nLastDsq);
```
Comment (furszy) [inline src/masternode.h:210]: how nodes that have this stored on disk will handle this change?.

**Review by furszy — COMMENTED** [review]
excellent add :), Code review ACK 58e3e34fe10a16c4e6fe759c31475b9e6603a013 .
Left some comments.

File: src/masternodeman.cpp [hunk 3]
```diff
@@ -482,6 +482,18 @@ CMasternode* CMasternodeMan::Find(const CPubKey& pubKeyMasternode)
     return nullptr;
 }
 
+void CMasternodeMan::CheckSpentCollaterals(const std::vector<CTransactionRef>& vtx)
+{
+    LOCK(cs);
+    for (const auto& tx : vtx) {
+        for (const auto& in : tx->vin) {
+            if (mapMasternodes.count(in.prevout)) {
```
[the remark this answers is not in the document]
Comment (random-zebra) [reply src/masternodeman.cpp:490]: yea. better

File: src/masternode.h [hunk 4]
```diff
@@ -230,6 +230,8 @@ class CMasternode : public CSignedMessage
         return lastPing.IsNull() ? false : now - lastPing.sigTime < seconds;
     }
 
+    void SetSpent() { activeState = CMasternode::MASTERNODE_VIN_SPENT; }
```
[the remark this answers is not in the document]
Comment (random-zebra) [reply src/masternode.h:220]: Yea, this is only temporary, as `activeState` is going away in the next PR, but will add the lock.

File: src/masternodeman.cpp [hunk 5]
```diff
@@ -210,7 +210,7 @@ bool CMasternodeMan::Add(CMasternode& mn)
 {
     LOCK(cs);
 
-    if (!mn.IsEnabled())
+    if (!mn.IsEnabled() && !mn.IsPreEnabled())
```
[the remark this answers is not in the document]
Comment (random-zebra) [reply src/masternodeman.cpp:213]: I don't see why we shouldn't be adding a pre-enabled masternode.
As for the question, well, it's covered by what you wrote as "Why this is working now" in #1886

File: src/qt/pivx/mnmodel.cpp (hunk 1 above)
Comment (random-zebra) [reply src/qt/pivx/mnmodel.cpp:34]: Yea, `MASTERNODE_MISSING` doesn't even exist anymore. I've put it `MASTERNODE_REMOVE` just as placeholder, it's going to go away in any case with the removal of `activeState`.  

But there is, actually, another bug, introduced here... setting `pmn->vin = CTxIn()`, which will leave the entry without collateral txid-txidn. Will revert that.

File: src/masternode.h (hunk 2 above)
Comment (random-zebra) [reply src/masternode.h:210]: Simple. At startup there will be an error deserializing (the log will show "Error reading mncache.dat..."), and the node will ask a new masternode list from its peers.

**Comment by random-zebra** [comment]
Updated

**Comment by random-zebra** [comment]
Great addition @furszy! :beer: 
This can be merged for me.

**Review by furszy — APPROVED** [review]
ACK 5d9148a

**Review by Fuzzbawls — APPROVED** [review]
utACK 5d9148aeefa4260eb6aad589054cc2c4af5f8906

**Pull request merged**